### PR TITLE
Adding an exception handler

### PR DESCRIPTION
**Pull request by zalox**
Hi! I've been tinkering with this library, and I'm a bit confused about the exceptions disappearing into some abyss while developing. I added a proposal for a handler, so that logging of exceptions on resources, for instance, is possible.

### CODE REVIEW

**Comment by eidheim** [comment]
This looks good, I have time Tomorrow morning to dive some more into it. Thank you!

File: server_http.hpp [hunk 1]
```diff
@@ -374,7 +383,9 @@ namespace SimpleWeb {
             try {
                 resource_function(response, request);
             }
-            catch(const std::exception&) {
+            catch(const std::exception&e) {
```
Comment (eidheim) [inline server_http.hpp]: Add a space here between `std::exception` and `&e`?

File: server_http.hpp [hunk 2]
```diff
@@ -90,7 +91,10 @@ namespace SimpleWeb {
                     remote_endpoint_address=socket.lowest_layer().remote_endpoint().address().to_string();
                     remote_endpoint_port=socket.lowest_layer().remote_endpoint().port();
                 }
-                catch(const std::exception&) {}
+                catch(const std::exception&e) {
```
Comment (eidheim) [inline server_http.hpp]: Add a space here between `std::exception` and `&e`?

File: server_http.hpp [hunk 3]
```diff
@@ -80,6 +80,7 @@ namespace SimpleWeb {
             std::string remote_endpoint_address;
             unsigned short remote_endpoint_port;
             
+            std::function<void(const std::exception&)> exception_handler;
```
Comment (eidheim) [inline server_http.hpp]: Maybe remove this, and move the try catch in Request::read_remote_endpoint_data to ServerBase:: read_request_and_content?

Comment (zalox) [inline server_http.hpp]: Why do we need the read_remote_endpoint_data?
```cpp
diff --git a/server_http.hpp b/server_http.hpp
index 70a86a2..440d36f 100644
--- a/server_http.hpp
+++ b/server_http.hpp
@@ -80,22 +80,10 @@ namespace SimpleWeb {
             std::string remote_endpoint_address;
             unsigned short remote_endpoint_port;
             
-            std::function<void(const std::exception&)> exception_handler;
         private:
             Request(): content(streambuf) {}
             
             boost::asio::streambuf streambuf;
-            
-            void read_remote_endpoint_data(socket_type& socket) {
-                try {
-                    remote_endpoint_address=socket.lowest_layer().remote_endpoint().address().to_string();
-                    remote_endpoint_port=socket.lowest_layer().remote_endpoint().port();
-                }
-                catch(const std::exception&e) {
-                    if(exception_handler)
-                       exception_handler(e);
-                }
-            }
         };
         
         class Config {
@@ -223,8 +211,14 @@ namespace SimpleWeb {
             //Create new streambuf (Request::streambuf) for async_read_until()
             //shared_ptr is used to pass temporary objects to the asynchronous functions
             std::shared_ptr<Request> request(new Request());
-            request->read_remote_endpoint_data(*socket);
 
+            try {
+                request->remote_endpoint_address=socket->lowest_layer().remote_endpoint().address().to_string();
+                request->remote_endpoint_port=socket->lowest_layer().remote_endpoint().port();
+            } catch(const std::exception &e){
+                if(exception_handler)
+                    exception_handler(e);
+            }
             //Set timeout on the following boost::asio::async-read or write function
             std::shared_ptr<boost::asio::deadline_timer> timer;
             if(timeout_request>0)
```
Is it to hide `boost::asio`  to implementors of the server?

Comment (eidheim) [inline server_http.hpp]: I guess we could change it to Request::address() (return ip adress) and Request::port() (return port) too, and let the user of the library handle any exceptions (though I can't see any cases where these functions can/should throw). Guess we should add noexcept on all the methods that does not throw at the same time. What do you think?

And yes, one point is to hide boost::asio from the user of the library. It's a pretty difficult library, and most people do not need to learn it.

Comment (eidheim) [inline server_http.hpp]: But I can do that later, your proposed change looks good. The problem with adding Request::address() and Request::port() is that the socket needs to be stored in the Request object somehow, and currently the Request objects can be stored longer than the socket object is alive.

**Pull request merged**